tokio/src/task/mod.rs

```diff
@@ -269,42 +264,65 @@ cfg_rt_util! {
     pub use task_local::LocalKey;
 }
 
+cfg_rt_threaded! {
+    mod stack;
+    pub(crate) use self::stack::TransferStack;
+}
+
 cfg_rt_core! {
+    /*
     /// Unit tests
     #[cfg(test)]
     mod tests;
+    */
 
     use std::future::Future;
     use std::marker::PhantomData;
     use std::ptr::NonNull;
     use std::{fmt, mem};
 
     /// An owned handle to the task, tracked by ref count
+    #[repr(transparent)]
     pub(crate) struct Task<S: 'static> {
         raw: RawTask,
         _p: PhantomData<S>,
     }
 
-    unsafe impl<S: ScheduleSendOnly + 'static> Send for Task<S> {}
+    unsafe impl<S: ScheduleSendOnly> Send for Task<S> {}
+    unsafe impl<S: ScheduleSendOnly> Sync for Task<S> {}
+
+    /// A task was notified
+    #[repr(transparent)]
+    pub(crate) struct Notified<S: 'static>(Task<S>);
+
+    unsafe impl<S: Schedule> Send for Notified<S> {}
+    unsafe impl<S: Schedule> Sync for Notified<S> {}
 
     /// Task result sent back
     pub(crate) type Result<T> = std::result::Result<T, JoinError>;
 
-    pub(crate) trait Schedule: Sized + 'static {
+    pub(crate) trait Schedule: Sync + Sized + 'static {
         /// Bind a task to the executor.
         ///
         /// Guaranteed to be called from the thread that called `poll` on the task.
-        fn bind(&self, task: &Task<Self>);
+        fn bind(task: &Task<Self>) -> Self;
 
         /// The task has completed work and is ready to be released. The scheduler
         /// is free to drop it whenever.
-        fn release(&self, task: Task<Self>);
-
-        /// The has been completed by the executor it was bound to.
-        fn release_local(&self, task: &Task<Self>);
+        ///
+        /// If the scheduler can immediately release the task, it should return
+        /// it as part of the function. This enables the task module to batch
+        /// the ref-dec with other options.
+        fn release(&self, task: Task<Self>) -> Option<Task<Self>>;
```

Comment:
Something is weird about this signature. If the release is happening from a remote thread, how can it possibly _own_ the `Task`? If it did, it could just release the `Task`. Doesn't this suggest that it's really more like a `&mut Task`? Otherwise, I think `Task` is poorly named, and should be called something like `TaskRef` or `TaskHandle` to indicate that the _owner_ is still the worker that manages the lifecycle of the task.

Reply:
I think that `Task` should be renamed to `task::Ref`. However, I opted to punt that. This PR is not the final PR!